File: point_to_offer/offer-ii-010.java
package point_to_offer;

import java.util.HashMap;
import java.util.Map;

class OfferII010 {
    /**
     * 前缀和 + HashTable记录前缀和
     * 检查 HashTable 中有没有 sum - k 这样的前缀和
     * 如果有，就ans += map.get(sum - k);
     * 如果没有，ans += 0
     * 然后将这个前缀和加入到 HashTable里面
     */
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        int sum = 0, ans = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            ans += map.getOrDefault(sum - k, 0);
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        return ans;
    }
}

File: point_to_offer/offer-ii-010.go
// 剑指 Offer II 010. 和为 k 的子数组 - Medium
// https://leetcode-cn.com/problems/QTMn0o/
package pointtooffer

// 这里想要用到滑动窗口的思想，但是看到题目的要求中数组的元素会有负数
// 这样的话就不符合滑动窗口的方案了，所以只能用到HashTable + 前缀和了
func subarraySum(nums []int, k int) int {
	dict := make(map[int]int)
	sum, ans := 0, 0
	for i := range nums {
		sum += nums[i]
		if _, ok := dict[sum-k]; ok {
			ans += dict[sum-k]
		}
		if _, ok := dict[sum]; ok {
			dict[sum] = dict[sum] + 1
		} else {
			dict[sum] = 1
		}
	}
	return ans
}
